File: MCP_Server/tools/actor_tools.py
from mcp.server.fastmcp import FastMCP
from client.unreal_client import UnrealClient


def register_actor_tools(mcp: FastMCP, client: UnrealClient):

    @mcp.tool()
    async def spawn_actor(
        class_name: str,
        name: str = "",
        location: list = None,
        rotation: list = None,
        scale: list = None
    ) -> str:
        params = {"className": class_name}
        if name:
            params["name"] = name
        if location:
            params["location"] = location
        if rotation:
            params["rotation"] = rotation
        if scale:
            params["scale"] = scale

        result = client.send_command("spawn_actor", params)
        if result.get("success"):
            return f"Spawned actor: {result.get('result', {})}"
        return f"Failed to spawn actor: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def destroy_actor(name: str) -> str:
        result = client.send_command("destroy_actor", {"name": name})
        if result.get("success"):
            return f"Destroyed actor: {name}"
        return f"Failed to destroy actor: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def set_actor_transform(
        name: str,
        location: list = None,
        rotation: list = None,
        scale: list = None
    ) -> str:
        params = {"name": name}
        if location:
            params["location"] = location
        if rotation:
            params["rotation"] = rotation
        if scale:
            params["scale"] = scale

        result = client.send_command("set_actor_transform", params)
        if result.get("success"):
            return f"Transform updated for {name}"
        return f"Failed: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def get_actor_list() -> str:
        result = client.send_command("get_actor_list", {})
        if result.get("success"):
            actors = result.get("result", [])
            return f"Actors in scene ({len(actors)}): {actors}"
        return f"Failed to get actor list: {result.get('error', 'Unknown error')}"

File: MCP_Server/src/tools/actor_tools.rs
use rmcp::model::*;
use rmcp::schemars;
use rmcp::server::tool;
use serde_json::json;
use crate::unreal_client::UnrealClient;
use std::sync::Arc;
use tokio::sync::Mutex;

pub struct ActorTools {
    client: Arc<Mutex<UnrealClient>>,
}

impl ActorTools {
    pub fn new(client: Arc<Mutex<UnrealClient>>) -> Self {
        Self { client }
    }

    #[tool(description = "Spawn an actor in the Unreal Engine scene")]
    async fn spawn_actor(
        &self,
        #[tool(param)]
        #[schemars(description = "Actor class name, e.g. 'StaticMeshActor', 'PointLight'")]
        class_name: String,
        #[tool(param)]
        #[schemars(description = "Optional actor name")]
        name: Option<String>,
        #[tool(param)]
        #[schemars(description = "Optional location [x, y, z]")]
        location: Option<Vec<f64>>,
        #[tool(param)]
        #[schemars(description = "Optional rotation [pitch, yaw, roll]")]
        rotation: Option<Vec<f64>>,
        #[tool(param)]
        #[schemars(description = "Optional scale [x, y, z]")]
        scale: Option<Vec<f64>>,
    ) -> Result<CallToolResult, McpError> {
        let mut params = json!({"className": class_name});
        if let Some(n) = name {
            params["name"] = json!(n);
        }
        if let Some(loc) = location {
            params["location"] = json!(loc);
        }
        if let Some(rot) = rotation {
            params["rotation"] = json!(rot);
        }
        if let Some(s) = scale {
            params["scale"] = json!(s);
        }

        let mut client = self.client.lock().await;
        match client.send_command("spawn_actor", params).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    format!("Spawned actor: {}", response["result"])
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }

    #[tool(description = "Destroy an actor by name")]
    async fn destroy_actor(
        &self,
        #[tool(param)]
        #[schemars(description = "Actor name to destroy")]
        name: String,
    ) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("destroy_actor", json!({"name": name})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    format!("Destroyed actor: {}", name)
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }

    #[tool(description = "Get list of all actors in the current scene")]
    async fn get_actor_list(&self) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("get_actor_list", json!({})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    format!("Actors: {}", response["result"])
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }
}
